File: CPU.cpp
//
// Created by alexs on 3/15/2025.
//

#include "CPU.h"

#include "Logger.h"
#include <bitset>

int CPU::exec(VirtualMachine& virtualMachine) {
    int commandAddress = virtualMachine.pc.toInteger();
    Word& command = virtualMachine.memory->readWord(commandAddress);

    Logger::debug("Executing command %.6s", command.word);

    bool isJump = false;
    if (command.equals("ADD000")) {
        Word& arg1 = virtualMachine.popFromStack();
        Word& arg2 = virtualMachine.popFromStack();
        Word result = arg1 + arg2;
        Logger::debug("ADD000: %.6s + %.6s = %.6s", arg1.word, arg2.word, result.word);

        virtualMachine.setZeroFlag(result);
        virtualMachine.setCarryFlag(arg1.toInteger() + arg2.toInteger() > result.toInteger());

        virtualMachine.pushToStack(result);
    } else if (command.equals("SUB000")) {
        Word& arg1 = virtualMachine.popFromStack();
        Word& arg2 = virtualMachine.popFromStack();
        Word result = arg1 - arg2;
        Logger::debug("SUB000: %.6s - %.6s = %.6s", arg1.word, arg2.word, result.word);

        virtualMachine.setZeroFlag(result);
        virtualMachine.setCarryFlag(arg1.toInteger() - arg2.toInteger() < result.toInteger());

        virtualMachine.pushToStack(result);
    } else if (command.equals("MUL000")) {
        Word& arg1 = virtualMachine.popFromStack();
        Word& arg2 = virtualMachine.popFromStack();
        Word result = arg1 * arg2;
        Logger::debug("MUL000: %.6s * %.6s = %.6s", arg1.word, arg2.word, result.word);
        virtualMachine.setZeroFlag(result);
        virtualMachine.pushToStack(result);
    } else if (command.equals("DIV000")) {
        Word& arg1 = virtualMachine.popFromStack();
        Word& arg2 = virtualMachine.popFromStack();
        Word result = arg1 / arg2;
        Logger::debug("DIV000: %.6s / %.6s = %.6s", arg1.word, arg2.word, result.word);
        virtualMachine.setZeroFlag(result);
        virtualMachine.pushToStack(result);
    } else if (command.equals("COMP00")) {
        Word& arg1 = virtualMachine.popFromStack();
        Word& arg2 = virtualMachine.popFromStack();
        Word result = arg1 - arg2;

        virtualMachine.setZeroFlag(result);
        virtualMachine.setCarryFlag(arg1.toInteger() - arg2.toInteger() < result.toInteger());
    } else if (command.equals("AND000")) {
        Word& arg1 = virtualMachine.popFromStack();
        Word& arg2 = virtualMachine.popFromStack();
        Word result = arg1 & arg2;
        Logger::debug("AND000: %.6s & %.6s = %.6s", arg1.word, arg2.word, result.word);
        virtualMachine.pushToStack(result);
    } else if (command.equals("OR0000")) {
        Word& arg1 = virtualMachine.popFromStack();
        Word& arg2 = virtualMachine.popFromStack();
        Word result = arg1 | arg2;
        Logger::debug("OR0000: %.6s | %.6s = %.6s", arg1.word, arg2.word, result.word);
        virtualMachine.pushToStack(result);
    } else if (command.equals("XOR000")) {
        Word& arg1 = virtualMachine.popFromStack();
        Word& arg2 = virtualMachine.popFromStack();
        Word result = arg1 ^ arg2;
        Logger::debug("XOR000: %.6s ^ %.6s = %.6s", arg1.word, arg2.word, result.word);
        virtualMachine.pushToStack(result);
    } else if (command.equals("JUMP00")) {
        Word& arg1 = virtualMachine.popFromStack();
        virtualMachine.pc = arg1;
        isJump = true;
        Logger::debug("JUMP00: %.6s", arg1.word);
    } else if (command.equals("JUMPB0")) {
        Word& arg1 = virtualMachine.popFromStack();
        if (virtualMachine.c & CF) {
            virtualMachine.pc = arg1;
            isJump = true;
            Logger::debug("JUMPB0: %.6s", arg1.word);
        } else {
            Logger::debug("JUMPB0: NO JUMP");
        }
    } else if (command.equals("JUMPBE")) {
        Word& arg1 = virtualMachine.popFromStack();
        if (virtualMachine.c & CF || virtualMachine.c & ZF) {
            virtualMachine.pc = arg1;
            isJump = true;
            Logger::debug("JUMPBE: %.6s", arg1.word);
        } else {
            Logger::debug("JUMPBE: NO JUMP");
        }
    } else if (command.equals("JUMPA0")) {
        Word& arg1 = virtualMachine.popFromStack();
        if (!(virtualMachine.c & CF) && !(virtualMachine.c & ZF)) {
            virtualMachine.pc = arg1;
            isJump = true;
            Logger::debug("JUMPA0: %.6s", arg1.word);
        } else {
            Logger::debug("JUMPA0: NO JUMP");
        }
    } else if (command.equals("JUMPE0")) {
        Word& arg1 = virtualMachine.popFromStack();
        if (virtualMachine.c & ZF) {
            virtualMachine.pc = arg1;
            isJump = true;
            Logger::debug("JUMPE0: %.6s", arg1.word);
        } else {
            Logger::debug("JUMPE0: NO JUMP");
        }
    } else if (command.equals("JUMPNE")) {
        Word& arg1 = virtualMachine.popFromStack();
        if (!(virtualMachine.c & ZF)) {
            virtualMachine.pc = arg1;
            isJump = true;
            Logger::debug("JUMPNE: %.6s", arg1.word);
        } else {
            Logger::debug("JUMPNE: NO JUMP");
        }
    } else if (command.startsWith("PU")) {
        std::string arg = command.substring(2); // Get the address from command inline args
        Word address("00" + arg); // Adjust the length of arg string
        Word& target = virtualMachine.memory->readWord(address.toInteger());
        virtualMachine.pushToStack(target);
        Logger::debug("PU: %.6s", target.word);
    } else if (command.startsWith("PO")) {
        std::string arg = command.substring(2); // Get the address from command inline args
        Word address("00" + arg); // Adjust the length of arg string
        Word& target = virtualMachine.popFromStack();
        virtualMachine.memory->writeWord(target, address.toInteger());
    } else if (command.equals("GETD00")) {
        signalSupervisorInterrupt(virtualMachine, SupervisorInterrupt::GETD);
        processInterrupt();
    } else if (command.equals("PRTW00")) {
        signalSupervisorInterrupt(virtualMachine, SupervisorInterrupt::PRTW);
        processInterrupt();
    } else if (command.equals("PRTS00")) {
        signalSupervisorInterrupt(virtualMachine, SupervisorInterrupt::PRTS);
        processInterrupt();
    } else if (command.equals("HALT00")) {
        signalSupervisorInterrupt(virtualMachine, SupervisorInterrupt::HALT);
        processInterrupt();
        return -1; // TODO: handle HALT in interrupt
    } else {
        Logger::debug("Unknown instruction: %.6s", command.word);
        return -1;
    }
    Logger::debug("Flags: %s", std::bitset<8>(virtualMachine.c).to_string().c_str());
    if (!isJump) {
        ++virtualMachine.pc;
    }
    return 0;
}

void CPU::signalInterrupt(VirtualMachine& virtualMachine, Byte interrupt, InterruptType type) {
    Interrupt newInterrupt{virtualMachine, interrupt, type};
    interrupts.push(newInterrupt);
}

void CPU::signalProgramInterrupt(VirtualMachine& virtualMachine, const ProgramInterrupt& interrupt) {
    pi = static_cast<Byte>(interrupt);
    signalInterrupt(virtualMachine, pi, InterruptType::PROGRAM);
}

void CPU::signalSupervisorInterrupt(VirtualMachine& virtualMachine, const SupervisorInterrupt& interrupt) {
    si = static_cast<Byte>(interrupt);
    signalInterrupt(virtualMachine, si, InterruptType::SUPERVISOR);
}

void CPU::processInterrupt() {
    if (interrupts.empty()) {
        return;
    }

    Interrupt interrupt = interrupts.front();
    interrupts.pop();

    if (interrupt.type == InterruptType::PROGRAM) {
        pi = interrupt.interrupt;
        processProgramInterrupt(interrupt.virtualMachine);

    } else if (interrupt.type == InterruptType::SUPERVISOR) {
        si = interrupt.interrupt;
        processSupervisorInterrupt(interrupt.virtualMachine);
    }
}

void CPU::processProgramInterrupt(VirtualMachine& virtualMachine) {
    ProgramInterrupt interrupt = static_cast<ProgramInterrupt>(pi);
    
    switch (interrupt) {
        case ProgramInterrupt::BADCODE:
        {
            Logger::debug("Bad code exception");
            break;
        }
        case ProgramInterrupt::SEGFAULT:
        {
            Logger::debug("Segmentation fault exception");
            break;
        }
        case ProgramInterrupt::OVERFLOW:
        {
            Logger::debug("Overflow exception");
            break;
        }
        case ProgramInterrupt::ZERODIV:
        {
            Logger::debug("Division by zero exception");
            break;
        }
        case ProgramInterrupt::BADNUM:
        {
            Logger::debug("Bad number exception");
            break;
        }
        default:
        {
            Logger::debug("Unknown program interrupt");
            break;
        }
    }
}

void CPU::processSupervisorInterrupt(VirtualMachine& virtualMachine) {
    SupervisorInterrupt interrupt = static_cast<SupervisorInterrupt>(si);
    
    switch (interrupt) {
        case SupervisorInterrupt::GETD:
        {
            Logger::debug("GETD supervisor interrupt");
            Word& address = virtualMachine.popFromStack();
            Word& size = virtualMachine.popFromStack();
            std::string buffer;
            std::getline(std::cin, buffer);

            for (int i = 0; i < size.toInteger(); ++i) {
                virtualMachine.memory->writeWord(Word(buffer.substr(i * WORD_SIZE, WORD_SIZE)), address.toInteger());
            }

            Logger::debug("GETD00: %.6s", address.word);
            break;
        }
        case SupervisorInterrupt::PRTW:
        {
            Logger::debug("PRTW supervisor interrupt");
            Word& arg = virtualMachine.popFromStack();
            std::cout << arg << std::endl;
            break;
        }
        case SupervisorInterrupt::PRTS:
        {
            Logger::debug("PRTS supervisor interrupt");
            break;
        }
        case SupervisorInterrupt::HALT:
        {
            Logger::debug("HALT supervisor interrupt");
            break;
        }
        default:
        {
            Logger::debug("Unknown supervisor interrupt");
            break;
        }
    }
}

File: src/CPU.java
import utils.Logger;

import java.util.*;

enum InterruptType {
    PROGRAM,
    SUPERVISOR
}

class Interrupt {
    public VirtualMachine virtualMachine;
    public byte interrupt;
    public InterruptType type;

    public Interrupt(VirtualMachine vm, byte interrupt, InterruptType type) {
        this.virtualMachine = vm;
        this.interrupt = interrupt;
        this.type = type;
    }
}

public class CPU {
    private Word ptr = new Word(0);
    private Word pc = new Word(0);
    private Word sp = new Word(65536);
    private byte c = 0;
    private byte pi = 0;
    private byte si = 0;
    private byte mode = 0;
    private byte ti = 0;

    private final Queue<Interrupt> interrupts = new LinkedList<>();

    public Word getPtr() {
        return ptr;
    }

    public void setPtr(Word word) {
        ptr = word;
    }

    public int exec(VirtualMachine vm) {
        int commandAddress = vm.pc.toInteger();
        Word command = vm.memory.readWord(commandAddress);

        Logger.debug(String.format("Executing command %.6s", command));

        boolean isJump = false;
        if (command.equals("ADD000")) {
            Word arg1 = vm.popFromStack();
            Word arg2 = vm.popFromStack();
            Word result = arg1.add(arg2);
            Logger.debug(String.format("ADD000: %.6s + %.6s = %.6s", arg1, arg2, result));

            vm.setZeroFlag(result);
            vm.setCarryFlag(arg1.toInteger() + arg2.toInteger() > result.toInteger());

            vm.pushToStack(result);
        } else if (command.equals("SUB000")) {
            Word arg1 = vm.popFromStack();
            Word arg2 = vm.popFromStack();
            Word result = arg1.subtract(arg2);
            Logger.debug(String.format("SUB000: %.6s - %.6s = %.6s", arg1, arg2, result));

            vm.setZeroFlag(result);
            vm.setCarryFlag(arg1.toInteger() - arg2.toInteger() < result.toInteger());

            vm.pushToStack(result);
        }
        else if (command.equals("MUL000")) {
            Word arg1 = vm.popFromStack();
            Word arg2 = vm.popFromStack();
            Word result = arg1.multiply(arg2);
            Logger.debug(String.format("MUL000: %.6s * %.6s = %.6s", arg1, arg2, result));
            vm.setZeroFlag(result);
            vm.pushToStack(result);
        }
        else if (command.equals("DIV000")) {
            Word arg1 = vm.popFromStack();
            Word arg2 = vm.popFromStack();
            if (arg2.toInteger() == 0) {
                signalProgramInterrupt(vm, ProgramInterrupt.ZERODIV);
                processInterrupt();
                return -1;
            }
            Word result = arg1.divide(arg2);
            Logger.debug(String.format("DIV000: %.6s / %.6s = %.6s", arg1, arg2, result));
            vm.setZeroFlag(result);
            vm.pushToStack(result);
        }
        else if (command.equals("COMP00")) {
            Word arg1 = vm.popFromStack();
            Word arg2 = vm.popFromStack();
            Word result = arg1.subtract(arg2);
            vm.setZeroFlag(result);
            vm.setCarryFlag(arg1.toInteger() - arg2.toInteger() < result.toInteger());
        }
        else if (command.equals("AND000")) {
            Word arg1 = vm.popFromStack();
            Word arg2 = vm.popFromStack();
            Word result = arg1.and(arg2);
            Logger.debug(String.format("AND000: %.6s & %.6s = %.6s", arg1, arg2, result));
            vm.pushToStack(result);
        }
        else if (command.equals("OR0000")) {
            Word arg1 = vm.popFromStack();
            Word arg2 = vm.popFromStack();
            Word result = arg1.or(arg2);
            Logger.debug(String.format("OR0000: %.6s | %.6s = %.6s", arg1, arg2, result));
            vm.pushToStack(result);
        }
        else if (command.equals("XOR000")) {
            Word arg1 = vm.popFromStack();
            Word arg2 = vm.popFromStack();
            Word result = arg1.xor(arg2);
            Logger.debug(String.format("XOR000: %.6s ^ %.6s = %.6s", arg1, arg2, result));
            vm.pushToStack(result);
        }
        else if (command.equals("JUMP00")) {
            Word arg1 = vm.popFromStack();
            vm.pc = arg1;
            isJump = true;
            Logger.debug(String.format("JUMP00: %.6s", arg1));
        }
        else if (command.equals("JUMPB0")) {
            Word arg1 = vm.popFromStack();
            if ((vm.c & VirtualMachine.CF) != 0) {
                vm.pc = arg1;
                isJump = true;
                Logger.debug(String.format("JUMPB0: %.6s", arg1));
            } else {
                Logger.debug("JUMPB0: NO JUMP");
            }
        }
        else if (command.equals("JUMPBE")) {
            Word arg1 = vm.popFromStack();
            if ((vm.c & VirtualMachine.CF) != 0 || (vm.c & VirtualMachine.ZF) != 0) {
                vm.pc = arg1;
                isJump = true;
                Logger.debug(String.format("JUMPBE: %.6s", arg1));
            } else {
                Logger.debug("JUMPBE: NO JUMP");
            }
        }
        else if (command.equals("JUMPA0")) {
            Word arg1 = vm.popFromStack();
            if ((vm.c & VirtualMachine.CF) == 0 && (vm.c & VirtualMachine.ZF) == 0) {
                vm.pc = arg1;
                isJump = true;
                Logger.debug(String.format("JUMPA0: %.6s", arg1));
            } else {
                Logger.debug("JUMPA0: NO JUMP");
            }
        }
        else if (command.equals("JUMPE0")) {
            Word arg1 = vm.popFromStack();
            if ((vm.c & VirtualMachine.ZF) != 0) {
                vm.pc = arg1;
                isJump = true;
                Logger.debug(String.format("JUMPE0: %.6s", arg1));
            } else {
                Logger.debug("JUMPE0: NO JUMP");
            }
        }
        else if (command.equals("JUMPNE")) {
            Word arg1 = vm.popFromStack();
            if ((vm.c & VirtualMachine.ZF) == 0) {
                vm.pc = arg1;
                isJump = true;
                Logger.debug(String.format("JUMPNE: %.6s", arg1));
            } else {
                Logger.debug("JUMPNE: NO JUMP");
            }
        }
        else if (command.startsWith("PU")) {
            String arg = command.substring(2);
            Word address = new Word("00" + arg);
            Word target = vm.memory.readWord(address.toInteger());
            vm.pushToStack(target);
            Logger.debug(String.format("PU: %.6s", target));
        }
        else if (command.startsWith("PO")) {
            String arg = command.substring(2);
            Word address = new Word("00" + arg);
            Word target = vm.popFromStack();
            vm.memory.writeWord(target, address.toInteger());
        }
        else if (command.equals("GETD00")) {
            signalSupervisorInterrupt(vm, SupervisorInterrupt.GETD);
            processInterrupt();
        }
        else if (command.equals("PRTW00")) {
            signalSupervisorInterrupt(vm, SupervisorInterrupt.PRTW);
            processInterrupt();
        }
        else if (command.equals("PRTS00")) {
            signalSupervisorInterrupt(vm, SupervisorInterrupt.PRTS);
            processInterrupt();
        }
        else if (command.equals("HALT00")) {
            signalSupervisorInterrupt(vm, SupervisorInterrupt.HALT);
            processInterrupt();
            return -1;
        } else {
            Logger.debug(String.format("Unknown instruction: %.6s", command));
            return -1;
        }

        Logger.debug("Flags: " + String.format("%8s", Integer.toBinaryString(vm.c & 0xFF)).replace(' ', '0'));

        if (!isJump) {
            vm.pc = vm.pc.add(new Word(1));
        }

        return 0;
    }

    private void signalInterrupt(VirtualMachine vm, byte interrupt, InterruptType type) {
        interrupts.add(new Interrupt(vm, interrupt, type));
    }

    public void signalProgramInterrupt(VirtualMachine vm, ProgramInterrupt interrupt) {
        this.pi = interrupt.value;
        signalInterrupt(vm, pi, InterruptType.PROGRAM);
    }

    public void signalSupervisorInterrupt(VirtualMachine vm, SupervisorInterrupt interrupt) {
        this.si = interrupt.value;
        signalInterrupt(vm, si, InterruptType.SUPERVISOR);
    }

    public void processInterrupt() {
        if (interrupts.isEmpty()) return;

        Interrupt interrupt = interrupts.poll();
        if (interrupt.type == InterruptType.PROGRAM) {
            this.pi = interrupt.interrupt;
            processProgramInterrupt(interrupt.virtualMachine);
        } else {
            this.si = interrupt.interrupt;
            processSupervisorInterrupt(interrupt.virtualMachine);
        }
    }

    private void processProgramInterrupt(VirtualMachine vm) {
        ProgramInterrupt interrupt = ProgramInterrupt.values()[pi];
        switch (interrupt) {
            case BADCODE -> Logger.debug("Bad code exception");
            case SEGFAULT -> Logger.debug("Segmentation fault exception");
            case OVERFLOW -> Logger.debug("Overflow exception");
            case ZERODIV -> Logger.debug("Division by zero exception");
            case BADNUM -> Logger.debug("Bad number exception");
            default -> Logger.debug("Unknown program interrupt");
        }
    }

    private void processSupervisorInterrupt(VirtualMachine vm) {
        SupervisorInterrupt interrupt = SupervisorInterrupt.values()[si];
        switch (interrupt) {
            case GETD -> {
                Logger.debug("GETD supervisor interrupt");
                Word size = vm.popFromStack();
                Word address = vm.popFromStack();
                Scanner scanner = new Scanner(System.in);
                String buffer = scanner.nextLine();

                for (int i = 0; i < size.toInteger(); ++i) {
                    String slice = buffer.substring(i * Constants.WORD_SIZE, (i + 1) * Constants.WORD_SIZE);
                    vm.memory.writeWord(new Word(slice), address.toInteger() + i);
                }
                Logger.debug("GETD00: " + address);
            }
            case PRTW -> {
                Logger.debug("PRTW supervisor interrupt");
                Word arg = vm.popFromStack();
                System.out.println(arg);
            }
            case PRTS -> Logger.debug("PRTS supervisor interrupt");
            case HALT -> Logger.debug("HALT supervisor interrupt");
            default -> Logger.debug("Unknown supervisor interrupt");
        }
    }
}
